runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java

```diff
@@ -0,0 +1,435 @@
+      }
+      // Set a timer to continue processing this element.
+      TimerInternals timerInternals = c.windowingInternals().timerInternals();
+      timerInternals.setTimer(
+          TimerInternals.TimerData.of(
+              stateNamespace,
+              timerInternals.currentProcessingTime().plus(cont.resumeDelay()),
+              TimeDomain.PROCESSING_TIME));
+    }
+
+    private DoFn<InputT, OutputT>.ProcessContext makeContext(
+        final ProcessContext baseContext,
+        final WindowedValue<InputT> element,
+        final TrackerT tracker,
+        final RestrictionT[] residualRestrictionHolder) {
+      return fn.new ProcessContext() {
+        private int numOutputs = 0;
+
+        public InputT element() {
+          return element.getValue();
+        }
+
+        public Instant timestamp() {
+          return element.getTimestamp();
+        }
+
+        public PaneInfo pane() {
+          return element.getPane();
+        }
+
+        public void output(OutputT output) {
+          baseContext
+              .windowingInternals()
+              .outputWindowedValue(
+                  output, element.getTimestamp(), element.getWindows(), element.getPane());
+          noteOutput();
+        }
+
+        public void outputWithTimestamp(OutputT output, Instant timestamp) {
+          baseContext
+              .windowingInternals()
+              .outputWindowedValue(output, timestamp, element.getWindows(), element.getPane());
+          noteOutput();
+        }
+
+        private void noteOutput() {
+          if (++numOutputs >= MAX_OUTPUTS_PER_BUNDLE) {
+            // Request a checkpoint. The fn *may* produce more output, but hopefully not too much.
+            residualRestrictionHolder[0] = tracker.checkpoint();
+          }
+        }
+
+        public <T> T sideInput(PCollectionView<T> view) {
+          return baseContext.sideInput(view);
+        }
+
+        public PipelineOptions getPipelineOptions() {
+          return baseContext.getPipelineOptions();
+        }
+
+        public <T> void sideOutput(TupleTag<T> tag, T output) {
+          // TODO: I'm not sure how to implement this correctly: there's no
+          // "internals.sideOutputWindowedValue".
+          throw new UnsupportedOperationException(
+              "Side outputs not yet supported by splittable DoFn");
+        }
+
+        public <T> void sideOutputWithTimestamp(TupleTag<T> tag, T output, Instant timestamp) {
+          // TODO: I'm not sure how to implement this correctly: there's no
+          // "internals.sideOutputWindowedValue".
+          throw new UnsupportedOperationException(
+              "Side outputs not yet supported by splittable DoFn");
+        }
+      };
+    }
+
+    /** Creates an {@link DoFn.ExtraContextFactory} that provides just the given tracker. */
+    private DoFn.ExtraContextFactory<InputT, OutputT> wrapTracker(final TrackerT tracker) {
+      return new DoFn.ExtraContextFactory<InputT, OutputT>() {
```

Comment:
I generally prefer "private static class" rather than creating these as anonymous inner classes, partly because it makes it clearer what is getting captured from the outside environment (here and above).